### ISSUE TITLE
Are you normalizing input data before passing to the model?

### ISSUE THREAD

**Issue by KushajveerSingh** [NNReasoning/What-Can-Neural-Networks-Reason-About#1]
I am trying to reproduce the results of your paper and I am having some difficulty.

In case of Maximum Value Difference you show that the MLP gets 9% accuracy. I was able to reproduce this result. But to get this result I did not normalize the input feature that is being passed to the MLP. 

When I normalize the input feature (divide the location by 20, divide value by 100, scale color to [0,1] range) I get 100% accuracy in a couple of iterations. 

In case of GNN, my model did not even converge when the data was not normalized. But when I normalize the data the GNN again got 100% accuracy in 1 epoch.

I understand it might not be possible to look at my code implementation due to time constraints, but in case it can help with the issue the code is available [here](https://github.com/KushajveerSingh/deep_learning/tree/master/graph_machine_learning/what_can_neural_networks_reason_about).

**Comment by jinglingli** [comment]
Hi Kushajveer,

Thanks for following our work! I have looked at your code and it seems that
you are using GCN rather than the GNN described in our paper for the GNN
model. GCN indeed could not obtain good generalization performance on our
proposed tasks because its structure is not that ''aligned'' with the task
algorithm (for more details, see the definition of ''algorithmic
alignment'' and theorem 3.6 in the paper). You need to use the GNN
described in equation (2.2) in order to reproduce the results.

Besides, I have several questions regarding your code:
1. How do you define classification after normalizing the input data? For
example, originally, we have 101 classes each corresponding to an integer
value of the maximum value difference (e.g. the kth class denotes the
maximum value is k, k>= 0). After normalization, I did not see the
remapping of the normalized value to a categorical value (e.g. did you map
0.01 to class 1, 0.02 to class 2, etc?). If not, it is not surprising that
you would get 100% accuracy as all data are all classified as the same
class (aka. class 0 or class 1 depending how you define the starting index).
2. I am not familiar with Trainer and I wonder if the acc reported by
trainer is training acc or test acc? Based on your code, I only see train
and validation datasets, and I did not see test dataset.

Meanwhile, feel free to modify based on our original repository to see if
you encounter such issues again. It would help the debug process as well
(we have implemented the data generation process and the MLP/GNN models as
well). Hope it helps!

Best,
Jingling

**Comment by KushajveerSingh** [comment]
Thanks for the reply. Indeed I had forgotten to map the classes after normalization. After reading the paper a few more times everything became more clear and what mistakes I was making. The concept of algorithm alignment gave me some new intuitions that I had not thought of before.

I had skipped the test set for simplicity as you had already provided hyperparameters.
